src/validation/code.rs

```diff
@@ -96,6 +107,35 @@ fn read_instructions(
             END => {
                 return Ok(());
             }
+            RETURN => {
+                let this_func_ty = &fn_types[type_idx_of_fn[this_function_idx]];
+
+                if value_stack.len() < this_func_ty.returns.valtypes.len() {
+                    return Err(Error::EndInvalidValueStack); // TODO(george-cosma): Is this the right error?
+                }
+
+                let ret_vals_start = value_stack.len() - this_func_ty.returns.valtypes.len();
+                let _remaining_locals = value_stack.drain(..ret_vals_start);
+
+                // TODO(george-cosma): There is a bug here. Because there is another op code after this reteurn (an
+                // `end`), I have to let this function consume it to work for the tests.
```

Comment:
As discussed previously, this is likely not a bug, since there is no evidence two `end`s  will ever be emitted by a compiler. 

The issue comes when a hand-made `wat` file which contains two `end`s is passed to the interpreter. In this case, after finishing with validation I propose we check that we reached the end of the function (all bytes were consumed).